iceoryx_dust/vocabulary/include/iox/detail/span.inl

```diff
@@ -0,0 +1,271 @@
+// Copyright (c) 2022 - 2023 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+#ifndef IOX_DUST_VOCABULARY_SPAN_INL
+#define IOX_DUST_VOCABULARY_SPAN_INL
+
+#include "iox/span.hpp"
+
+namespace iox
+{
+template <typename Container>
+inline constexpr auto data(Container& container) -> decltype(container.data())
+{
+    return container.data();
+}
+
+template <typename Container>
+inline constexpr auto data(const Container& container) -> decltype(container.data())
+{
+    return container.data();
+}
+
+template <typename T, uint64_t N>
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-c-arrays, hicpp-avoid-c-arrays)
+inline constexpr T* data(T (&array)[N]) noexcept
+{
+    return array;
+}
+
+template <typename T, uint64_t N, template <typename, uint64_t> class Buffer>
+inline constexpr T* data(UninitializedArray<T, N, Buffer>& uninit_array) noexcept
+{
+    return uninit_array.begin();
+}
+
+template <typename T, uint64_t Extent>
+template <typename It>
+inline constexpr span<T, Extent>::span(It first, uint64_t count) noexcept
+    : span_storage_t(count)
+    , m_data(detail::to_address(first))
+{
+    assert(Extent == DYNAMIC_EXTENT || Extent == count);
+}
+
+template <typename T, uint64_t Extent>
+template <typename It, typename End, typename>
+inline constexpr span<T, Extent>::span(It begin, End end) noexcept
+    : span(begin, static_cast<uint64_t>(end - begin))
+{
+    static_assert(begin <= end, "begin shall not be smaller than end");
+}
+
+template <typename T, uint64_t Extent>
+template <uint64_t N, typename>
+inline constexpr span<T, Extent>::span(
+    T (&array)[N]) noexcept // NOLINT(cppcoreguidelines-avoid-c-arrays, hicpp-avoid-c-arrays)
+    : span(iox::data(array), N)
+{
+}
+
+template <typename T, uint64_t Extent>
+template <typename U, uint64_t N, template <typename, uint64_t> class Buffer, typename>
+inline constexpr span<T, Extent>::span(iox::UninitializedArray<U, N, Buffer>& array) noexcept
+    : span(iox::data(array), N)
+{
+}
+
+template <typename T, uint64_t Extent>
+template <typename U, uint64_t N, template <typename, uint64_t> class Buffer, typename>
+inline constexpr span<T, Extent>::span(const iox::UninitializedArray<U, N, Buffer>& array) noexcept
+    : span(iox::data(array), N)
+{
+}
```

Comment:
Since `data` and `size` are already specialized for `T (&array)[N]` and `UninitializedArray`, are these c`tors really necessary?

Reply:
Yes, they are. Otherwise such code will not compile `span<const int> const_sut(array)`.